cnd/src/db/save_message.rs

```diff
@@ -0,0 +1,353 @@
+use crate::{
+    db::{
+        custom_sql_types::{Text, U32},
+        new_types::{DecimalU256, EthereumAddress, Satoshis},
+        schema::{self, *},
+        Error, Sqlite,
+    },
+    swap_protocols::{
+        ledger::{Bitcoin, Ethereum},
+        rfc003::{Accept, Decline, Request, SecretHash},
+        HashFunction, SwapId,
+    },
+};
+use diesel::RunQueryDsl;
+
+pub trait SaveMessage<M> {
+    fn save_message(&self, message: M) -> Result<(), Error>;
+}
+
+pub trait SaveRfc003Messages:
+    SaveMessage<Request<Bitcoin, Ethereum, bitcoin::Amount, ethereum_support::EtherQuantity>>
+    + SaveMessage<Request<Bitcoin, Ethereum, bitcoin::Amount, ethereum_support::Erc20Token>>
+    + SaveMessage<Request<Ethereum, Bitcoin, ethereum_support::EtherQuantity, bitcoin::Amount>>
+    + SaveMessage<Request<Ethereum, Bitcoin, ethereum_support::Erc20Token, bitcoin::Amount>>
+    + SaveMessage<Accept<Bitcoin, Ethereum>>
+    + SaveMessage<Accept<Ethereum, Bitcoin>>
+    + SaveMessage<Decline>
+{
+}
+
+macro_rules! impl_save_message {
+    (
+        fn save_message($conn:ident : SqliteConnection, $var:ident : $message:ty) ->
+        $ret:ty
+        $body:block
+    ) => {
+        impl SaveMessage<$message> for Sqlite {
+            fn save_message(&self, $var: $message) -> $ret {
+                let $conn = self.connect().unwrap();
+
+                $body
+            }
+        }
+    };
+}
+
+#[derive(Insertable, Debug, Clone)]
+#[table_name = "rfc003_bitcoin_ethereum_bitcoin_ether_request_messages"]
+struct InsertableBitcoinEthereumBitcoinEtherRequestMessage {
+    swap_id: Text<SwapId>,
+    bitcoin_network: Text<bitcoin::Network>,
+    ethereum_chain_id: U32,
+    bitcoin_amount: Text<Satoshis>,
+    ether_amount: Text<DecimalU256>,
+    hash_function: Text<HashFunction>,
+    bitcoin_refund_identity: Text<bitcoin::PublicKey>,
+    ethereum_redeem_identity: Text<EthereumAddress>,
+    bitcoin_expiry: U32,
+    ethereum_expiry: U32,
+    secret_hash: Text<SecretHash>,
+}
+
+impl_save_message! {
+    fn save_message(connection: SqliteConnection, message: Request<Bitcoin, Ethereum, bitcoin::Amount, ethereum_support::EtherQuantity>) -> Result<(), Error> {
+        let Request {
+            id,
```

Comment:
The use of `id` here not to represent the table `id` but to represent the table `swap_id` only reinforce my conviction that we should remove `id` from the table (and only have `swap_id`).